xds/internal/balancer/priority/balancer_priority.go

```diff
@@ -0,0 +1,354 @@
+		b.priorityInitTimer = timerW
+		timerW.timer = time.AfterFunc(defaultPriorityInitTimeout, func() {
+			b.mu.Lock()
+			defer b.mu.Unlock()
+			if timerW.stopped {
+				return
+			}
+			b.priorityInitTimer = nil
+			// Switch to the next priority if there's any.
+			if pNext := priority + 1; pNext < len(b.priorities) {
+				nameNext := b.priorities[pNext]
+				if childNext, ok := b.children[nameNext]; ok {
+					b.switchToChild(childNext, pNext)
+					childNext.sendUpdate()
+				}
+			}
+		})
+	}
+}
+
+// handleChildStateUpdate start/close priorities based on the connectivity
+// state.
+func (b *priorityBalancer) handleChildStateUpdate(childName string, s balancer.State) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.done.HasFired() {
+		return
+	}
+
+	priority, ok := b.childToPriority[childName]
+	if !ok {
+		b.logger.Warningf("priority: received picker update with unknown child %v", childName)
+		return
+	}
+
+	if b.childInUse == "" {
+		b.logger.Warningf("priority: no child is in use when picker update is received")
+		return
+	}
+
+	// priorityInUse is higher than this priority.
+	if b.priorityInUse < priority {
+		// Lower priorities should all be closed, this is an unexpected update.
+		// Can happen if the child policy sends an update after we tell it to
+		// close.
+		b.logger.Warningf("priority: received picker update from priority %v,  lower than priority in use %v", priority, b.priorityInUse)
+		return
+	}
+
+	// Update state in child. The updated picker will be sent to parent later if
+	// necessary.
+	child, ok := b.children[childName]
+	if !ok {
+		b.logger.Errorf("priority: child balancer not found for child %v, priority %v", childName, priority)
+		return
+	}
+	oldState := child.state.ConnectivityState
+	child.state = s
+
+	switch s.ConnectivityState {
+	case connectivity.Ready:
+		b.handlePriorityWithNewStateReady(child, priority)
+	case connectivity.TransientFailure:
+		b.handlePriorityWithNewStateTransientFailure(child, priority)
+	case connectivity.Connecting:
+		b.handlePriorityWithNewStateConnecting(child, priority, oldState)
+	default:
+		// New state is Idle, should never happen. Don't forward.
+	}
+}
+
+// handlePriorityWithNewStateReady handles state Ready from a higher or equal
+// priority.
+//
+// An update with state Ready:
+// - If it's from higher priority:
+//   - Switch to this priority
+//   - Forward the update
+// - If it's from priorityInUse:
+//   - Forward only
+//
+// Caller must make sure priorityInUse is not higher than priority.
+//
+// Caller must hold mu.
+func (b *priorityBalancer) handlePriorityWithNewStateReady(child *childBalancer, priority int) {
+	// If one priority higher or equal to priorityInUse goes Ready, stop the
+	// init timer. If update is from higher than priorityInUse, priorityInUse
+	// will be closed, and the init timer will become useless.
+	b.stopPriorityInitTimer()
+
+	// priorityInUse is lower than this priority, switch to this.
+	if b.priorityInUse > priority {
```

Comment:
Do we really need this check here? `switchToChild` does handle the case where `priorityInUse` is the same as `priority`, by doing nothing.